server/sonar-server/src/main/java/org/sonar/server/computation/period/PeriodsRepository.java

```diff
@@ -0,0 +1,187 @@
+
+  private static final int NUMBER_OF_PERIODS = 5;
+
+  private final DbClient dbClient;
+  private final Settings settings;
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodFinder periodFinder;
+  private final BatchReportReader batchReportReader;
+
+  private List<Period> periods = new ArrayList<>();
+
+  public PeriodsRepository(DbClient dbClient, Settings settings, TreeRootHolder treeRootHolder, PeriodFinder periodFinder, BatchReportReader batchReportReader) {
+    this.dbClient = dbClient;
+    this.settings = settings;
+    this.treeRootHolder = treeRootHolder;
+    this.periodFinder = periodFinder;
+    this.batchReportReader = batchReportReader;
+  }
+
+  public List<Period> getPeriods(){
+   if (periods.isEmpty()) {
+      initPeriods();
+    }
+    return periods;
+  }
+
+  private void initPeriods() {
+    DbSession session = dbClient.openSession(false);
+    try {
+      Component project = treeRootHolder.getRoot();
+      ComponentDto projectDto = dbClient.componentDao().selectNullableByKey(session, project.getKey());
+      // No project on first analysis, no period
+      if (projectDto != null) {
+        BatchReport.Component batchProject = batchReportReader.readComponent(project.getRef());
+        PeriodResolver periodResolver = new PeriodResolver(session, projectDto.getId(), batchReportReader.readMetadata().getAnalysisDate(), batchProject.getVersion(),
+          // TODO qualifier will be different for Views
+          Qualifiers.PROJECT);
+
+        for (int index = 1; index <= NUMBER_OF_PERIODS; index++) {
+          Period period = periodResolver.resolve(index);
+          // SONAR-4700 Add a past snapshot only if it exists
+          if (period != null) {
+            periods.add(period.setIndex(index));
+            LOG.debug(period.toString());
+          }
+        }
+      }
+    } finally {
+      session.close();
+    }
+  }
+
+  private class PeriodResolver {
+
+    private final DbSession session;
+    private final long projectId;
+    private final long analysisDate;
+    private final String currentVersion;
+    private final String qualifier;
+
+    public PeriodResolver(DbSession session, long projectId, long analysisDate, String currentVersion, String qualifier) {
+      this.session = session;
+      this.projectId = projectId;
+      this.analysisDate = analysisDate;
+      this.currentVersion = currentVersion;
+      this.qualifier = qualifier;
+    }
+
+    @CheckForNull
+    private Period resolve(int index) {
+      String propertyValue = getPropertyValue(qualifier, settings, index);
+      Period period = resolve(index, propertyValue);
+      if (period == null && StringUtils.isNotBlank(propertyValue)) {
+        LOG.debug("Property " + CoreProperties.TIMEMACHINE_PERIOD_PREFIX + index + " is not valid: " + propertyValue);
+      }
+      return period;
+    }
+
+    @CheckForNull
+    private Period resolve(int index, String property) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "resolve" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)